src/hdmf_zarr/zarr_utils.py

```diff
@@ -0,0 +1,289 @@
+
+    @docval({'name': 'dataset', 'type': (np.ndarray, ZarrArray, Array), 'doc': 'the Zarr file lazily evaluate'},
+            {'name': 'io', 'type': 'ZarrIO', 'doc': 'the IO object that was used to read the underlying dataset'},
+            {'name': 'types', 'type': (list, tuple),
+             'doc': 'the list/tuple of reference types'})
+    def __init__(self, **kwargs):
+        types = popargs('types', kwargs)
+        super().__init__(**kwargs)
+        self.__refgetters = dict()
+        for i, t in enumerate(types):
+            # if t is RegionReference: # not yet supported
+            #     self.__refgetters[i] = self.__get_regref
+            if t == DatasetBuilder.OBJECT_REF_TYPE:
+                # breakpoint()
+                self.__refgetters[i] = self._get_ref
+            elif t is str:
+                # we need this for when we read compound data types
+                # that have unicode sub-dtypes since Zarrpy does not
+                # store UTF-8 in compound dtypes
+                self.__refgetters[i] = self._get_utf
+        self.__types = types
+        tmp = list()
+        for i in range(len(self.dataset.dtype)):
+            sub = self.dataset.dtype[i]
+            if sub.metadata:
+                if 'vlen' in sub.metadata:
+                    t = sub.metadata['vlen']
+                    if t is str:
+                        tmp.append('utf')
+                    elif t is bytes:
+                        tmp.append('ascii')
+                elif 'ref' in sub.metadata:
+                    t = sub.metadata['ref']
+                    if t is Reference:
+                        tmp.append('object')
+                    elif t is RegionReference:
+                        tmp.append('region')
+            else:
+                tmp.append(sub.type.__name__)
+        self.__dtype = tmp
+
+    @property
+    def types(self):
+        return self.__types
+
+    @property
+    def dtype(self):
+        return self.__dtype
+
+    def __getitem__(self, arg):
+        rows = copy(super().__getitem__(arg))
+        if np.issubdtype(type(arg), np.integer):
+            self.__swap_refs(rows)
+        else:
+            for row in rows:
+                self.__swap_refs(row)
+        return rows
+
+    def __swap_refs(self, row):
+        for i in self.__refgetters:
+            getref = self.__refgetters[i]
+            row[i] = getref(row[i])
+
+    def _get_utf(self, string):
+        """
+        Decode a dataset element to unicode
+        """
+        return string.decode('utf-8') if isinstance(string, bytes) else string
+
+    def __get_regref(self, ref):
+        obj = self._get_ref(ref)
+        return obj[ref]
+
+    def resolve(self, manager):
+        return self[0:len(self)]
+
+    def __iter__(self):
+        for i in range(len(self)):
+            yield self[i]
+
+
+class AbstractZarrReferenceDataset(DatasetOfReferences):
+
```

Comment:
Add docstring for these classes.

For the classes for "Region" references, I would add a note that ZarrIO does not yet fully support region references and that the classes are, therefore, currently unused.